nixos/modules/virtualisation/azure-image.nix

```diff
@@ -71,11 +79,38 @@ in
       format = "raw";
 
       bootSize = "${toString cfg.bootSize}M";
-      partitionTableType = if cfg.vmGeneration == "v2" then "efi" else "legacy";
+      partitionTableType = if (cfg.vmGeneration == "v2") then "efi" else "legacy";
 
-      inherit (cfg) contents;
+      inherit (cfg) contents label;
       inherit (config.virtualisation) diskSize;
       inherit config lib pkgs;
     };
+
+    boot.growPartition = true;
+    boot.loader.grub = rec {
+      efiSupport = (cfg.vmGeneration == "v2");
+      device = if efiSupport then "nodev" else "/dev/sda";
+      efiInstallAsRemovable = efiSupport;
+      font = null;
+      splashImage = null;
+      extraConfig = lib.mkIf (!efiSupport) ''
```

Comment:
The below needs explanatory comments, you might use the ones I provided in https://github.com/NixOS/nixpkgs/pull/333508#discussion_r1857498306.